sdk/metrics-testing/src/main/java/io/opentelemetry/sdk/testing/exporter/InMemoryMetricReader.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.testing.exporter;
+
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.metrics.data.AggregationTemporality;
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import io.opentelemetry.sdk.metrics.export.MetricProducer;
+import io.opentelemetry.sdk.metrics.export.MetricReader;
+import io.opentelemetry.sdk.metrics.export.MetricReaderFactory;
+import java.util.Collection;
+import java.util.Collections;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link MetricReader} implementation that can be used to test OpenTelemetry integration.
+ *
+ * <p>Can be created using {@code InMemoryMetricReader.create()}
+ *
+ * <p>Example usage:
+ *
+ * <pre><code>
+ * public class InMemoryMetricReaderExample {
+ *   private final InMemoryMetricReader reader = InMemoryMetricReader.create();
+ *   private final SdkMeterProvider sdkMeterProvider = SdkMeterProvider.builder().registerMetricReader(reader).build();
+ *   private final Meter meter = sdkMeterProvider.get("example");
+ *   private final LongCounter metricCallCount = meter.counterBuilder("num_collects");
+ *
+ *   public void printMetrics() {
+ *     metricCallCount.add(1);
+ *     System.out.println(reader.collectAllMetrics());
+ *   }
+ *
+ *   public static void main(String[] args) {
+ *     InMemoryMetricReaderExample example = new InMemoryMetricReaderExample();
+ *     example.printMetrics();
+ *   }
+ * }
+ * </code></pre>
+ */
+public class InMemoryMetricReader implements MetricReader, MetricReaderFactory {
+  private final AggregationTemporality preferred;
+  @Nullable private volatile MetricProducer metricProducer;
+
+  /** Returns a new {@link InMemoryMetricReader}. */
+  public static InMemoryMetricReader create() {
+    return new InMemoryMetricReader(AggregationTemporality.CUMULATIVE);
+  }
+
+  /** Creates a new {@link InMemoryMetricReader} that prefers DELTA aggregation. */
+  public static InMemoryMetricReader createDelta() {
+    return new InMemoryMetricReader(AggregationTemporality.DELTA);
+  }
+
+  private InMemoryMetricReader(AggregationTemporality preferred) {
+    this.preferred = preferred;
+  }
+
+  /** Returns all metrics accumulated since the last call. */
+  public Collection<MetricData> collectAllMetrics() {
+    MetricProducer metricProducer = this.metricProducer;
+    if (metricProducer != null) {
+      return metricProducer.collectAllMetrics();
+    }
+    return Collections.emptyList();
+  }
```

Comment:
Not a big deal but we may as well add the test case for this case (collect before apply)

Reply:
Done. 

While I was at it I noticed that there were two tests classes for this that were largely duplicative, so I merged them. Also noticed that the `shutdown()` method didn't actually do anything to prevent future calls to `collectAllMetrics()`, so I adjusted its behavior to set `metricProducer = null`, returning it to its state pre `apply(MetricProducer)`.